### PR TITLE
fix: when overriding paymaster set fields to 0x initially

### PR DESCRIPTION
**Pull request by moldy530**
# Pull Request Checklist

- [ ] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [ ] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [ ] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [ ] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [ ] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:write`)
- [ ] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

## PR-Codex overview
This PR introduces a new `ClientMiddleware` type and refactors the `runMiddlewareStack` function to handle different middleware scenarios.

### Detailed summary
- Added `ClientMiddleware` type
- Refactored `runMiddlewareStack` to handle `dummyPaymasterAndData`
- Updated logic for `dummyPaymasterAndData` with conditional fields based on overrides

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts [hunk 1]
```diff
@@ -73,11 +74,20 @@ export async function _runMiddlewareStack<
...
+    ClientMiddleware<TContext>,
+    "dummyPaymasterAndData" | "paymasterAndData"
+  > = bypassPaymasterAndData(overrides)
     ? {
-        dummyPaymasterAndData: noopMiddleware,
+        dummyPaymasterAndData: async (uo, overrides) => ({
+          ...uo,
+          ...("paymasterAndData" in overrides
+            ? { paymasterAndData: "0x" }
+            : { paymaster: "0x" }),
```
[the remark this answers is not in the document]
Comment (moldy530) [reply packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts:89]: oh good to know! ok cool lemme fix that

**Review by avasisht23 — DISMISSED** [review]

File: packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts [hunk 2]
```diff
@@ -73,11 +74,20 @@ export async function _runMiddlewareStack<
...
+    ClientMiddleware<TContext>,
+    "dummyPaymasterAndData" | "paymasterAndData"
+  > = bypassPaymasterAndData(overrides)
     ? {
-        dummyPaymasterAndData: noopMiddleware,
+        dummyPaymasterAndData: async (uo, overrides) => ({
+          ...uo,
+          ...("paymasterAndData" in overrides
+            ? { paymasterAndData: "0x" }
+            : { paymaster: undefined }),
```
Comment (dphilipson) [inline packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts:89]: Actually, really sorry, but can we set all four paymaster fields to undefined in this case (`paymaster`, `paymasterData`, `paymasterVerificationGasLimit` and `paymasterPostOpGasLimit`)? A strict reading of the spec doesn't allow any of the other three fields to be set if `paymaster` is not. Although Rundler and other common bundlers accept this, ideally the SDK should adhere to the spec.

Comment (moldy530) [reply packages/core/src/actions/smartAccount/internal/runMiddlewareStack.ts:89]: nothing else should be set at this point actually, so I can just leave it out

**Review by dphilipson — APPROVED** [review]
Thank you!

**Pull request merged**